MSB3371: parallel multi-TFM build collides on shared MetalamaSourceGenerator.touch file

## Problem

PR #1478 (issue #670) changed `MetalamaSourceGeneratorTouchFile` from `$(IntermediateOutputPath)` (per-TFM, e.g. `obj\Debug\net8.0-windows\`) to `$(BaseIntermediateOutputPath)` (the shared `obj\` root). The reasoning was that the touch file was occasionally observed empty in design-time scenarios.

The unintended consequence is that **every TFM of a multi-targeted project now writes to the same touch file**. When MSBuild builds two TFMs in parallel (default in `dotnet pack` / `dotnet build` for SDK-style multi-TFM projects), the parallel `Touch` tasks collide on the file lock and one of them fails with **MSB3371**.

Observed in CI builds of `Metalama.Patterns.Wpf` (a multi-targeted WPF project):

```
error MSB3371: The file "...\Metalama.Patterns.Wpf\obj\MetalamaSourceGenerator.touch" cannot be created.
The process cannot access the file because it is being used by another process.
[...\Metalama.Patterns.Wpf.csproj::TargetFramework=net8.0-windows]
```

## Proposed fix

Revert the touch file path back to `$(IntermediateOutputPath)` (TFM-specific). The original empty-string symptom from #670 was caused by the property being undefined when the .NET SDK has not yet resolved (e.g. early evaluation / IDE scenarios). Address that root cause directly: when `IntermediateOutputPath` is undefined, simply do not write the touch file at all (skip the `<Touch>` task and the `AdditionalFiles` include). The Roslyn analyzer / source generator already tolerates a missing path on the read side (`SourceGeneratorTouchFile` is nullable in `IProjectOptions`).

## Repro

```
git clone <repo>
dotnet pack Metalama.Patterns\Metalama.Patterns.sln -c Debug
```

(reproducible on CI; locally the race is timing-dependent)

## Acceptance

- Multi-TFM projects build in parallel without MSB3371 on the touch file.
- The empty-string regression from #670 does not return: when the SDK is unresolved, `IProjectOptions.SourceGeneratorTouchFile` is null/empty and the rest of the pipeline still works.
- Add a regression test for the multi-TFM parallel-build case (or a unit test confirming the targets file produces a TFM-specific path).

— Claude for @gfraiteur

--- a/Metalama.Framework/src/Metalama.Framework.Package/build/Metalama.Framework.targets
+++ b/Metalama.Framework/src/Metalama.Framework.Package/build/Metalama.Framework.targets
@@ -78,21 +78,25 @@
         </PropertyGroup>
     </Target>
 
-    <Target Name="ExportMetalamaFrameworkProperties" BeforeTargets="PrepareForBuild">
-        <Error Text="The BaseIntermediateOutputPath property is not defined." Condition="'$(BaseIntermediateOutputPath)'==''"/>
+    <!-- Both touch files live under $(IntermediateOutputPath) so they are TFM-specific. Multi-targeted
+         projects build their TFMs in parallel; sharing a single file across TFMs (as we briefly did under
+         $(BaseIntermediateOutputPath) for the source-generator touch file) causes MSB3371 file-lock
+         collisions. See https://github.com/metalama/Metalama/issues/1597.
+         If $(IntermediateOutputPath) is undefined (e.g. when the .NET SDK has not yet resolved), the
+         path properties stay empty and CreateMetalamaTouchFiles is skipped — IProjectOptions
+         tolerates a null SourceGeneratorTouchFile/BuildTouchFile. -->
+    <Target Name="ExportMetalamaFrameworkProperties" BeforeTargets="PrepareForBuild" Condition="'$(IntermediateOutputPath)'!=''">
         <PropertyGroup>
             <MetalamaBuildTouchFile>$([System.IO.Path]::Combine($(MSBuildProjectDirectory),$(IntermediateOutputPath),'MetalamaBuild.touch'))</MetalamaBuildTouchFile>
-            <MetalamaSourceGeneratorTouchFile>$([System.IO.Path]::Combine($(MSBuildProjectDirectory),$(BaseIntermediateOutputPath),'MetalamaSourceGenerator.touch'))</MetalamaSourceGeneratorTouchFile>
+            <MetalamaSourceGeneratorTouchFile>$([System.IO.Path]::Combine($(MSBuildProjectDirectory),$(IntermediateOutputPath),'MetalamaSourceGenerator.touch'))</MetalamaSourceGeneratorTouchFile>
         </PropertyGroup>
     </Target>
 
-    <Target Name="CreateMetalamaTouchFiles" BeforeTargets="BeforeCompile">
+    <Target Name="CreateMetalamaTouchFiles" BeforeTargets="BeforeCompile" Condition="'$(MetalamaBuildTouchFile)'!='' AND '$(MetalamaSourceGeneratorTouchFile)'!=''">
         <!-- We create a touch file every time that the file is absent. This file is used to signal
              the design-time Metalama pipeline that a build has started. When the design-time pipeline needs this information again, it will
              delete the file. It should do that only when the source code has changed (even if it is not saved). We don't overwrite an existing
-             file because this would break incremental build.
-             We use Touch with AlwaysCreate=true which handles concurrent access gracefully when
-             multiple TFMs build in parallel since they share the same obj directory. -->
+             file because this would break incremental build. -->
         <Touch Files="$(MetalamaBuildTouchFile)" AlwaysCreate="true"/>
         <Touch Files="$(MetalamaSourceGeneratorTouchFile)" AlwaysCreate="true"/>
         <ItemGroup>

--- a/Metalama.Framework/src/tests/Metalama.Framework.Tests.UnitTests/Options/MSBuildProjectOptionsTests.cs
+++ b/Metalama.Framework/src/tests/Metalama.Framework.Tests.UnitTests/Options/MSBuildProjectOptionsTests.cs
@@ -54,6 +54,19 @@ public sealed class MSBuildProjectOptionsTests
         Assert.Equal( touchFilePath, options.SourceGeneratorTouchFile );
     }
 
+    [Fact]
+    public void MissingSourceGeneratorTouchFile_ReturnsNull()
+    {
+        // When IntermediateOutputPath is undefined (e.g. SDK not yet resolved), the targets file skips
+        // ExportMetalamaFrameworkProperties and the property is never exposed to the compiler. See #1597.
+        var source = new DictionaryOptionsSource( new Dictionary<string, string>() );
+
+        var options = new TestableMSBuildProjectOptions( source );
+
+        Assert.Null( options.SourceGeneratorTouchFile );
+        Assert.Null( options.BuildTouchFile );
+    }
+
     private sealed class TestableMSBuildProjectOptions : MSBuildProjectOptions
     {
         public TestableMSBuildProjectOptions( IProjectOptionsSource source ) : base( source ) { }

